packages/x-date-pickers/src/internals/hooks/usePicker/usePickerValue.ts

```diff
@@ -255,58 +141,61 @@ export const usePickerValue = <
-      isControlled,
-      closeOnSelect,
-    };
-
-    const shouldPublish = shouldPublishValue(updaterParams);
-    const shouldCommit = shouldCommitValue(updaterParams);
-    const shouldClose = shouldClosePicker(updaterParams);
+  const setValue = useEventCallback((newValue: TValue, options?: SetValueActionOptions<TError>) => {
+    const {
+      changeImportance = 'accept',
+      skipPublicationIfPristine = false,
+      validationError,
+      shortcut,
+    } = options ?? {};
+
+    let shouldPublish: boolean;
+    let shouldCommit: boolean;
+    if (!skipPublicationIfPristine && !isControlled && !dateState.hasBeenModifiedSinceMount) {
+      // If the value is not controlled and the value has never been modified before,
+      // Then clicking on any value (including the one equal to `defaultValue`) should call `onChange` and `onAccept`
+      shouldPublish = true;
+      shouldCommit = changeImportance === 'accept';
+    } else {
+      shouldPublish = !valueManager.areValuesEqual(utils, newValue, dateState.lastPublishedValue);
+      shouldCommit =
+        changeImportance === 'accept' &&
+        !valueManager.areValuesEqual(utils, newValue, dateState.lastCommittedValue);
+    }
 
     setDateState((prev) => ({
       ...prev,
-      draft: action.value,
-      lastPublishedValue: shouldPublish ? action.value : prev.lastPublishedValue,
-      lastCommittedValue: shouldCommit ? action.value : prev.lastCommittedValue,
+      draft: newValue,
+      lastPublishedValue: shouldPublish ? newValue : prev.lastPublishedValue,
+      lastCommittedValue: shouldCommit ? newValue : prev.lastCommittedValue,
       hasBeenModifiedSinceMount: true,
     }));
 
     let cachedContext: PickerChangeHandlerContext<TError> | null = null;
     const getContext = (): PickerChangeHandlerContext<TError> => {
       if (!cachedContext) {
-        const validationError =
-          action.name === 'setExplicitValue' && action.options.validationError != null
-            ? action.options.validationError
-            : getValidationErrorForNewValue(action.value);
-
         cachedContext = {
-          validationError,
+          validationError:
+            validationError == null ? getValidationErrorForNewValue(newValue) : validationError,
         };
 
-        if (action.name === 'setExplicitValue') {
-          if (action.options.shortcut) {
-            cachedContext.shortcut = action.options.shortcut;
-          }
+        if (shortcut) {
+          cachedContext.shortcut = shortcut;
         }
       }
 
       return cachedContext;
     };
 
     if (shouldPublish) {
-      handleValueChange(action.value, getContext());
+      handleValueChange(newValue, getContext());
     }
 
     if (shouldCommit && onAccept) {
-      onAccept(action.value, getContext());
+      onAccept(newValue, getContext());
     }
 
-    if (shouldClose) {
+    if (changeImportance === 'accept') {
```

Comment:
Could you merge the master to see if everything still checks out after https://github.com/mui/mui-x/pull/15944 has been merged? 🤔

Reply:
Done
I'll double check once the CI has passed or failed

Reply:
Thanks. All looks solid. 👍